File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/fbnetwork/elk/ELKLayoutProvider.java
package org.fbme.ide.richediting.adapters.fbnetwork.elk;

import com.intellij.openapi.util.Pair;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.eclipse.elk.alg.layered.options.LayeredMetaDataProvider;
import org.eclipse.elk.core.IGraphLayoutEngine;
import org.eclipse.elk.core.RecursiveGraphLayoutEngine;
import org.eclipse.elk.core.data.ILayoutMetaDataProvider;
import org.eclipse.elk.core.data.LayoutMetaDataService;
import org.eclipse.elk.core.options.CoreOptions;
import org.eclipse.elk.core.util.NullElkProgressMonitor;
import org.eclipse.elk.graph.*;
import org.eclipse.elk.graph.util.ElkGraphUtil;
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionCursor;
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionPath;
import org.fbme.ide.richediting.viewmodel.*;
import org.fbme.lib.iec61499.fbnetwork.ConnectionPath;
import org.fbme.scenes.controllers.SceneViewpoint;
import org.fbme.scenes.controllers.components.ComponentController;
import org.fbme.scenes.controllers.components.ComponentSynchronizer;
import org.fbme.scenes.controllers.components.ComponentsFacility;
import org.fbme.scenes.controllers.diagram.*;

import java.awt.*;
import java.util.List;
import java.util.*;

import static org.junit.Assert.assertTrue;

public class ELKLayoutProvider {
//    private static final Logger LOG = LogManager.getLogger(ELKLayoutProvider.class);

    private final DiagramFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, Point> diagramFacility;
    private final ComponentsFacility<NetworkComponentView, Point> componentsFacility;
    private final ConnectionsFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, FBConnectionCursor, FBConnectionPath> connectionsFacility;
    private final DiagramController<NetworkComponentView, NetworkPortView, NetworkConnectionView> diagramController;
    private final ComponentSynchronizer<NetworkComponentView, Point> componentSynchronizer;
    private final ConnectionPathSynchronizer<NetworkConnectionView, FBConnectionPath> connectionSynchronizer;
    private final SceneViewpoint viewpoint;

    private final IGraphLayoutEngine layoutEngine = new RecursiveGraphLayoutEngine();
    private final List<ILayoutMetaDataProvider> layoutProviders = Arrays.asList(new CoreOptions(), new LayeredMetaDataProvider());
    private final ELKProperties layoutPropertiesProvider = new ELKProperties();

    public ELKLayoutProvider(
            DiagramFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, Point> diagramFacility,
            ComponentsFacility<NetworkComponentView, Point> componentsFacility,
            ConnectionsFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, FBConnectionCursor, FBConnectionPath> connectionsFacility,
            SceneViewpoint viewpoint
    ) {
        this.diagramFacility = diagramFacility;
        this.componentsFacility = componentsFacility;
        this.connectionsFacility = connectionsFacility;
        this.diagramController = diagramFacility.getDiagramController();
        this.componentSynchronizer = componentsFacility.getComponentSynchronizer();
        this.connectionSynchronizer = connectionsFacility.getConnectionSynchronizer();
        this.viewpoint = viewpoint;
    }

    public void relayout() {
        try {
            LayoutMetaDataService.getInstance().registerLayoutMetaDataProviders(layoutProviders.toArray(new ILayoutMetaDataProvider[0]));
            List<Pair<NetworkComponentView, ElkNode>> nodes = new ArrayList<>();
            Map<NetworkPortView, ElkPort> mapViewPort = new HashMap<>();
            List<Pair<NetworkConnectionView, ElkEdge>> edges = new ArrayList<>();
            List<Pair<InterfaceEndpointView, ElkPort>> endpoints = new ArrayList<>();
            ElkNode root = createElkGraph(nodes, edges, endpoints, mapViewPort);
            layoutEngine.layout(root, new NullElkProgressMonitor());
            applyLayout(nodes, endpoints, edges);
        } catch (Throwable t) {
//            LOG.error("error when relayout");
        }
    }

    private ElkNode createElkGraph(List<Pair<NetworkComponentView, ElkNode>> nodes, List<Pair<NetworkConnectionView, ElkEdge>> edges, List<Pair<InterfaceEndpointView, ElkPort>> endpoints, Map<NetworkPortView, ElkPort> mapViewPort) {
        ElkNode root = ElkGraphUtil.createGraph();
        ElkNode node = ElkGraphUtil.createNode(root);
        node.setLocation(0.0, 0.0);
        layoutPropertiesProvider.setRootProperties(node);

        processComponents(node, nodes, endpoints, mapViewPort);
        processConnections(node, mapViewPort, edges);

        return root;
    }

    private void processConnections(ElkNode parent, Map<NetworkPortView, ElkPort> mapViewPort, List<Pair<NetworkConnectionView, ElkEdge>> edges) {
        for (NetworkConnectionView connection : diagramController.getConnections()) {
            ElkPort sourceElkPort = mapViewPort.get(diagramController.getSource(connection));
            ElkPort targetElkPort = mapViewPort.get(diagramController.getTarget(connection));

            ElkEdge elkEdge = ElkGraphUtil.createSimpleEdge(sourceElkPort, targetElkPort);
            elkEdge.setContainingNode(parent);
            edges.add(new Pair<>(connection, elkEdge));
        }
    }

    private void processComponents(ElkNode parent, List<Pair<NetworkComponentView, ElkNode>> nodes, List<Pair<InterfaceEndpointView, ElkPort>> endpoints, Map<NetworkPortView, ElkPort> mapViewPort) {
        for (NetworkComponentView component : diagramController.getComponents()) {
            ComponentController<Point> componentController = componentsFacility.getController(component);
            Point componentPosition = componentsFacility.getModelForm(component);
            Rectangle componentBounds = componentController.getBounds(componentPosition);
            int x = viewpoint.translateFromEditorX(componentBounds.x);
            int y = viewpoint.translateFromEditorY(componentBounds.y);
            int width = viewpoint.fromEditorDimension(componentBounds.width);
            int height = viewpoint.fromEditorDimension(componentBounds.height);

            if (component instanceof InterfaceEndpointView) {
                InterfaceEndpointView endpoint = (InterfaceEndpointView) component;
                ElkPort port = ElkGraphUtil.createPort(parent);
                port.setDimensions(width, height);
                layoutPropertiesProvider.setPortProperties(port, endpoint.isSource());
                mapViewPort.put(endpoint, port);
                endpoints.add(new Pair<>(endpoint, port));
                continue;
            }

            ElkNode node = ElkGraphUtil.createNode(parent);
            node.setLocation(x, y);
            node.setDimensions(width, height);
            layoutPropertiesProvider.setNodeProperties(node);

            for (NetworkPortView componentPort : diagramController.getPorts(component)) {
                PortController componentPortController = diagramController.getPortController(componentPort);
                Rectangle componentPortBounds = componentPortController.getBounds();
                int portX = viewpoint.translateFromEditorX(componentPortBounds.x);
                int portY = viewpoint.translateFromEditorY(componentPortBounds.y);
                int portWidth = viewpoint.fromEditorDimension(componentPortBounds.width);
                int portHeight = viewpoint.fromEditorDimension(componentPortBounds.height);

                ElkPort port = ElkGraphUtil.createPort(node);
                port.setLocation(portX - x, portY - y);
                port.setDimensions(portWidth, portHeight);
                if (componentPort instanceof FunctionBlockPortView) {
                    layoutPropertiesProvider.setPortProperties(port, ((FunctionBlockPortView) componentPort).isSource());
                }
                mapViewPort.put(componentPort, port);
            }
            nodes.add(new Pair<>(component, node));
        }
    }

    private void applyLayout(List<Pair<NetworkComponentView, ElkNode>> nodes, List<Pair<InterfaceEndpointView, ElkPort>> endpoints, List<Pair<NetworkConnectionView, ElkEdge>> edges) {
        applyNodeLayout(nodes);
        applyEndpointLayout(endpoints);
        applyEdgeLayout(edges);
    }

    private void applyEndpointLayout(List<Pair<InterfaceEndpointView, ElkPort>> endpoints) {
        for (Pair<InterfaceEndpointView, ElkPort> endpoint : endpoints) {
            ElkPort elkPort = endpoint.second;
            int x = viewpoint.translateToEditorX((int) elkPort.getX());
            int y = viewpoint.translateToEditorY((int) elkPort.getY());

            componentSynchronizer.setForm(endpoint.first, new Point(x, y));
        }
    }

    private void applyNodeLayout(List<Pair<NetworkComponentView, ElkNode>> nodes) {
        for (Pair<NetworkComponentView, ElkNode> node : nodes) {
            ElkNode elkNode = node.second;
            int x = viewpoint.translateToEditorX((int) elkNode.getX());
            int y = viewpoint.translateToEditorY((int) elkNode.getY());

            componentSynchronizer.setForm(node.first, new Point(x, y));
        }
    }

    private void applyEdgeLayout(List<Pair<NetworkConnectionView, ElkEdge>> edges) {
        for (Pair<NetworkConnectionView, ElkEdge> edge : edges) {
            ElkEdge elkEdge = edge.second;
            List<Point> points = getPointsFromEdge(elkEdge);
            assertTrue(points.size() >= 2 && points.size() % 2 == 0);
            FBConnectionPath connectionPath = getConnectionPathFromPoints(points);

            connectionSynchronizer.setPath(edge.first, connectionPath);
        }
    }

    private ArrayList<Point> getPointsFromEdge(ElkEdge elkEdge) {
        ArrayList<Point> points = new ArrayList<>();
        for (ElkEdgeSection section : elkEdge.getSections()) {
            int startX = viewpoint.translateToEditorX((int) section.getStartX());
            int startY = viewpoint.translateToEditorY((int) section.getStartY());
            points.add(new Point(startX, startY));
            for (ElkBendPoint bendPoint : section.getBendPoints()) {
                int x = viewpoint.translateToEditorX((int) bendPoint.getX());
                int y = viewpoint.translateToEditorY((int) bendPoint.getY());
                points.add(new Point(x, y));
            }
            int endX = viewpoint.translateToEditorX((int) section.getEndX());
            int endY = viewpoint.translateToEditorY((int) section.getEndY());
            points.add(new Point(endX, endY));
        }
        return points;
    }

    private FBConnectionPath getConnectionPathFromPoints(List<Point> points) {
        int count = points.size();
        assertTrue(count >= 2 && count % 2 == 0);
        if (count > 6) {
            return new FBConnectionPath(
                    points.get(0),
                    points.get(points.size() - 1),
                    points.subList(1, points.size() - 1)
            );
        }

        int x1 = 0, y = 0, x2 = 0;
        ConnectionPath.Kind pathKind = ConnectionPath.Kind.Straight;
        if (count > 2) {
            x1 = points.get(1).x;
            pathKind = ConnectionPath.Kind.TwoAngles;
            if (count > 4) {
                x2 = points.get(count - 2).x;
                y = points.get(2).y;
                pathKind = ConnectionPath.Kind.FourAngles;
            }
        }
        return new FBConnectionPath(points.get(0), points.get(count - 1), pathKind, x1, y, x2);
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/fbnetwork/elk/ELKLayoutProvider.kt
package org.fbme.ide.richediting.adapters.fbnetwork.elk

import org.eclipse.elk.alg.layered.options.LayeredMetaDataProvider
import org.eclipse.elk.core.IGraphLayoutEngine
import org.eclipse.elk.core.RecursiveGraphLayoutEngine
import org.eclipse.elk.core.data.LayoutMetaDataService
import org.eclipse.elk.core.options.CoreOptions
import org.eclipse.elk.core.util.NullElkProgressMonitor
import org.eclipse.elk.graph.ElkEdge
import org.eclipse.elk.graph.ElkNode
import org.eclipse.elk.graph.ElkPort
import org.eclipse.elk.graph.util.ElkGraphUtil
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionCursor
import org.fbme.ide.richediting.adapters.fbnetwork.FBConnectionPath
import org.fbme.ide.richediting.adapters.fbnetwork.InlineValueController.Companion.INNER_BORDER_PADDING
import org.fbme.ide.richediting.adapters.fbnetwork.InlineValueController.Companion.OPPOSITE_PORT_PADDING
import org.fbme.ide.richediting.inspections.NetworkInspectionsFacility
import org.fbme.ide.richediting.viewmodel.*
import org.fbme.lib.iec61499.fbnetwork.ConnectionPath
import org.fbme.scenes.controllers.SceneViewpoint
import org.fbme.scenes.controllers.components.ComponentsFacility
import org.fbme.scenes.controllers.diagram.ConnectionsFacility
import org.fbme.scenes.controllers.diagram.DiagramFacility
import org.junit.Assert
import java.awt.Point
import java.awt.Rectangle

class ELKLayoutProvider(
    diagramFacility: DiagramFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, Point>,
    private val componentsFacility: ComponentsFacility<NetworkComponentView, Point>,
    connectionsFacility: ConnectionsFacility<NetworkComponentView, NetworkPortView, NetworkConnectionView, FBConnectionCursor, FBConnectionPath>,
    private val inspectionsFacility: NetworkInspectionsFacility,
    private val viewpoint: SceneViewpoint
) {
    private val diagramController = diagramFacility.diagramController
    private val componentSynchronizer = componentsFacility.componentSynchronizer
    private val connectionSynchronizer = connectionsFacility.connectionSynchronizer
    private val layoutEngine: IGraphLayoutEngine = RecursiveGraphLayoutEngine()
    private val layoutProviders = listOf(CoreOptions(), LayeredMetaDataProvider())
    private val layoutPropertiesProvider = ELKProperties()

    fun relayout() {
        LayoutMetaDataService.getInstance().registerLayoutMetaDataProviders(*layoutProviders.toTypedArray())
        val nodesMap: MutableMap<FunctionBlockView, ElkNode> = mutableMapOf()
        val mapViewPort: MutableMap<NetworkPortView, ElkPort> = HashMap()
        val edgesMap: MutableMap<NetworkConnectionView, ElkEdge> = mutableMapOf()
        val endpointsMap: MutableMap<InterfaceEndpointView, ElkPort> = mutableMapOf()
        val root = createElkGraph(nodesMap, edgesMap, endpointsMap, mapViewPort)
        layoutEngine.layout(root, NullElkProgressMonitor())
        applyLayout(nodesMap, endpointsMap, edgesMap)
    }

    private fun createElkGraph(
        nodesMap: MutableMap<FunctionBlockView, ElkNode>,
        edgesMap: MutableMap<NetworkConnectionView, ElkEdge>,
        endpointsMap: MutableMap<InterfaceEndpointView, ElkPort>,
        mapViewPort: MutableMap<NetworkPortView, ElkPort>
    ): ElkNode {
        val root = ElkGraphUtil.createGraph()
        val node = ElkGraphUtil.createNode(root)
        node.setLocation(0.0, 0.0)
        layoutPropertiesProvider.setRootProperties(node)
        processComponents(node, nodesMap, endpointsMap, mapViewPort)
        processConnections(node, mapViewPort, edgesMap)
        return root
    }

    private fun processConnections(
        parent: ElkNode,
        mapViewPort: Map<NetworkPortView, ElkPort>,
        edges: MutableMap<NetworkConnectionView, ElkEdge>
    ) {
        val connections = diagramController.connections
        for (connection in connections) {
            val sourceElkPort = mapViewPort[diagramController.getSource(connection)]
            val targetElkPort = mapViewPort[diagramController.getTarget(connection)]
            val elkEdge = ElkGraphUtil.createSimpleEdge(sourceElkPort, targetElkPort)
            elkEdge.containingNode = parent
            edges[connection] = elkEdge
        }
    }

    private fun processComponents(
        parent: ElkNode,
        nodesMap: MutableMap<FunctionBlockView, ElkNode>,
        endpointsMap: MutableMap<InterfaceEndpointView, ElkPort>,
        mapViewPort: MutableMap<NetworkPortView, ElkPort>
    ) {
        val components = diagramController.components
        val endpoints = components.filterIsInstance<InterfaceEndpointView>()
        val functionBlocks = components.filterIsInstance<FunctionBlockView>()
        val inlineValues = components.filterIsInstance<InlineValueView>()

        for (endpoint in endpoints) {
            val bounds = getBounds(endpoint)
            val port = ElkGraphUtil.createPort(parent)
            port.setDimensions(bounds.getWidth(), bounds.getHeight())
            layoutPropertiesProvider.setPortProperties(port, endpoint.isSource)
            mapViewPort[endpoint] = port
            endpointsMap[endpoint] = port
        }

        for (functionBlock in functionBlocks) {
            val bounds = getBounds(functionBlock)
            val elkNode = ElkGraphUtil.createNode(parent)
            elkNode.setLocation(bounds.getX(), bounds.getY())
            elkNode.setDimensions(bounds.getWidth(), bounds.getHeight())
            layoutPropertiesProvider.setNodeProperties(elkNode)
            val ports = diagramController.getPorts(functionBlock) as Set<FunctionBlockPortView>
            for (port in ports) {
                val portBounds = getBounds(port)
                val elkPort = ElkGraphUtil.createPort(elkNode)
                elkPort.setLocation((portBounds.x - bounds.x).toDouble(), (portBounds.y - bounds.y).toDouble())
                elkPort.setDimensions(portBounds.getWidth(), portBounds.getHeight())
                layoutPropertiesProvider.setPortProperties(elkPort, port.isSource)
                mapViewPort[port] = elkPort
            }
            nodesMap[functionBlock] = elkNode
        }

        for (inlineValue in inlineValues) {
            val bounds = getBounds(inlineValue)
            val functionBlock = inlineValue.opposite.component
            val elkNode = nodesMap[functionBlock]
            val port = ElkGraphUtil.createPort(elkNode)
            val padding = OPPOSITE_PORT_PADDING + 2 * INNER_BORDER_PADDING
            port.setDimensions(bounds.getWidth() + padding, bounds.getHeight())
            val fbBounds = getBounds(functionBlock)
            port.setLocation(bounds.getX() - fbBounds.getX(), bounds.getY() - fbBounds.getY())
            layoutPropertiesProvider.setPortProperties(port, true)
            mapViewPort[inlineValue] = port
        }
    }

    private fun getBounds(port: FunctionBlockPortView): Rectangle {
        val componentPortController = diagramController.getPortController(port)
        val componentPortBounds = componentPortController.bounds
        return viewpoint.fromEditor(componentPortBounds)
    }

    private fun getBounds(endpoint: InterfaceEndpointView): Rectangle {
        val componentController = componentsFacility.getController(endpoint)
        val componentPosition = componentsFacility.getModelForm(endpoint)
        val componentBounds = componentController.getBounds(componentPosition)
        return viewpoint.fromEditor(componentBounds)
    }

    private fun getBounds(functionBlock: FunctionBlockView): Rectangle {
        val componentController = componentsFacility.getController(functionBlock)
        val componentPosition = componentsFacility.getModelForm(functionBlock)
        val componentBounds = componentController.getBounds(componentPosition)
        return viewpoint.fromEditor(componentBounds)
    }

    private fun getBounds(inlineValue: InlineValueView): Rectangle {
        val componentController = inspectionsFacility.componentProvider(inlineValue)
        val componentPosition = inspectionsFacility.layoutModel.getActivePosition(inlineValue)
        val componentBounds = componentController.getBounds(componentPosition)
        return viewpoint.fromEditor(componentBounds)
    }

    private fun applyLayout(
        nodes: Map<FunctionBlockView, ElkNode>,
        endpoints: Map<InterfaceEndpointView, ElkPort>,
        edges: Map<NetworkConnectionView, ElkEdge>
    ) {
        applyNodeLayout(nodes)
        applyEndpointLayout(endpoints)
        applyEdgeLayout(edges)
    }

    private fun applyEndpointLayout(endpointsMap: Map<InterfaceEndpointView, ElkPort>) {
        for ((endpoint, elkPort) in endpointsMap) {
            val x = viewpoint.translateToEditorX(elkPort.x.toInt())
            val y = viewpoint.translateToEditorY(elkPort.y.toInt())
            componentSynchronizer.setForm(endpoint, Point(x, y))
        }
    }

    private fun applyNodeLayout(nodesMap: Map<FunctionBlockView, ElkNode>) {
        for ((functionBlock, elkNode) in nodesMap) {
            val x = viewpoint.translateToEditorX(elkNode.x.toInt())
            val y = viewpoint.translateToEditorY(elkNode.y.toInt())
            componentSynchronizer.setForm(functionBlock, Point(x, y))
        }
    }

    private fun applyEdgeLayout(edgesMap: Map<NetworkConnectionView, ElkEdge>) {
        for ((connection, elkEdge) in edgesMap) {
            val points: List<Point> = getPointsFromEdge(elkEdge)
            assert(points.size >= 2 && points.size % 2 == 0)
            val connectionPath = getConnectionPathFromPoints(points)
            connectionSynchronizer.setPath(connection, connectionPath)
        }
    }

    private fun getPointsFromEdge(elkEdge: ElkEdge): ArrayList<Point> {
        val points = ArrayList<Point>()
        for (section in elkEdge.sections) {
            val startX = viewpoint.translateToEditorX(section.startX.toInt())
            val startY = viewpoint.translateToEditorY(section.startY.toInt())
            points.add(Point(startX, startY))
            for (bendPoint in section.bendPoints) {
                val x = viewpoint.translateToEditorX(bendPoint.x.toInt())
                val y = viewpoint.translateToEditorY(bendPoint.y.toInt())
                points.add(Point(x, y))
            }
            val endX = viewpoint.translateToEditorX(section.endX.toInt())
            val endY = viewpoint.translateToEditorY(section.endY.toInt())
            points.add(Point(endX, endY))
        }
        return points
    }

    private fun getConnectionPathFromPoints(points: List<Point>): FBConnectionPath {
        val count = points.size
        Assert.assertTrue(count >= 2 && count % 2 == 0)
        if (count > 6) {
            return FBConnectionPath(
                points[0],
                points[points.size - 1],
                points.subList(1, points.size - 1).toMutableList()
            )
        }
        var x1 = 0
        var y = 0
        var x2 = 0
        var pathKind = ConnectionPath.Kind.Straight
        if (count > 2) {
            x1 = points[1].x
            pathKind = ConnectionPath.Kind.TwoAngles
            if (count > 4) {
                x2 = points[count - 2].x
                y = points[2].y
                pathKind = ConnectionPath.Kind.FourAngles
            }
        }
        return FBConnectionPath(points[0], points[count - 1], pathKind, x1, y, x2)
    }
}
